Spanner/src/Connection/Grpc.php

```diff
@@ -1092,4 +1167,71 @@ private function getDatabaseAdminClient()
 
         return $this->databaseAdminClient;
     }
+
+    /**
+     * Allow lazy instantiation of the spanner client.
+     * If routing is enabled, a SpannerClient is returned for a specific instanceName
+     * with a binding in the connection of a specific endpoint.
+     * If the endpoint list is not requested, a request is first made for this instanceName.
+     *
+     * @param string $instanceName Optional. Full name of instance.
+     * @return SpannerClient
+     */
+    private function getSpannerClient($instanceName = null)
+    {
+        if (!$this->spannerClient) {
+            $this->spannerClient = $this->constructGapic(SpannerClient::class, $this->grpcConfig);
+        }
+
+        if ($this->enableResourceCaching) {
+            if (isset($this->spannerClients[$instanceName])) {
+                return $this->spannerClients[$instanceName];
+            } elseif (!isEmpty($instanceName) && !isset($this->endpointUris[$instanceName])) {
+                $parsed = InstanceAdminClient::parseName($instanceName);
+                $projectId = $parsed['project'];
+
+                try {
+                    $instanceInfo = $this->getInstance([
+                        'projectId' => $projectId,
+                        'name' => $instanceName,
+                        'fieldMask' => ['endpointUris']
+                    ]);
+                    $this->endpointUris[$instanceName] = $instanceInfo['endpointUris'];
```

Comment:
Perhaps we can check if `$instanceInfo['endpointUris']` is not empty, then we won't need to check for `null` on line 1200.

Reply:
If `$instanceInfo['endpointUris']` came empty, I believe an entry should be inserted int `$this->spannerClients_[$instanceName => defaultClient]` in order not to call getInstance on every call.

Just putting here as an idea/note/suggestion
clear cache logic for a particular instance can be exposed to the user (later)

Reply:
The isset construct `$this->endpointUris [$instanceName]` will verify the fact of the request for this $instanceName, because some value will be written there. If the request was not made isset, it will return false, because `$instanceName` in the `$this->endpointUris` array will not be found. As a result of repeated requests will not happen.
The endpoint method returns one of uri or null. This is a different operation. because the presence of `$instanceName` in the `$this->endpointUris` array does not mean success, because `$this->endpointUris[$instanceName]` may contain an empty list.

I think it’s more correct to leave the implementation proposed by me.